pkg/driver/node.go

```diff
@@ -0,0 +1,566 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package driver
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"time"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/kubernetes-csi/csi-lib-iscsi/iscsi"
+	"github.com/openebs/jiva-csi/pkg/kubernetes/client"
+	"github.com/openebs/jiva-csi/pkg/request"
+	jv "github.com/openebs/jiva-operator/pkg/apis/openebs/v1alpha1"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	// FSTypeExt2 represents the ext2 filesystem type
+	FSTypeExt2 = "ext2"
+	// FSTypeExt3 represents the ext3 filesystem type
+	FSTypeExt3 = "ext3"
+	// FSTypeExt4 represents the ext4 filesystem type
+	FSTypeExt4 = "ext4"
+	// FSTypeXfs represents te xfs filesystem type
+	FSTypeXfs = "xfs"
+
+	defaultFsType = FSTypeExt4
+
+	MaxRetryCount = 10
+)
+
+var (
+	ValidFSTypes = []string{FSTypeExt2, FSTypeExt3, FSTypeExt4, FSTypeXfs}
+)
+
+var (
+	// nodeCaps represents the capability of node service.
+	nodeCaps = []csi.NodeServiceCapability_RPC_Type{
+		csi.NodeServiceCapability_RPC_STAGE_UNSTAGE_VOLUME,
+	}
+)
+
+// node is the server implementation
+// for CSI NodeServer
+type node struct {
+	client           *client.Client
+	driver           *CSIDriver
+	mounter          *NodeMounter
+	volumeTransition *request.Transition
+}
+
+// NewNode returns a new instance
+// of CSI NodeServer
+func NewNode(d *CSIDriver, cli *client.Client) csi.NodeServer {
+	return &node{
+		client:           cli,
+		driver:           d,
+		mounter:          newNodeMounter(),
+		volumeTransition: request.NewTransition(),
+	}
+}
+
+func (ns *node) attachDisk(instance *jv.JivaVolume) (string, error) {
+	connector := iscsi.Connector{
+		VolumeName:    instance.Name,
+		TargetIqn:     instance.Spec.ISCSISpec.Iqn,
+		Lun:           instance.Spec.ISCSISpec.Lun,
+		Interface:     instance.Spec.ISCSISpec.ISCSIInterface,
+		TargetPortals: instance.Spec.ISCSISpec.TargetPortals,
+		Discovery:     true,
+	}
+
+	logrus.Debugf("NodeStageVolume: attach disk with config: {%+v}", connector)
+	devicePath, err := iscsi.Connect(connector)
+	if err != nil {
+		return "", err
+	}
+
+	if devicePath == "" {
+		return "", fmt.Errorf("connect reported success, but no path returned")
+	}
+	return devicePath, err
+}
+
+func (ns *node) waitForVolumeToBeReady(volID string) (*jv.JivaVolume, error) {
+	var retry int32
+	for {
+		if err := ns.client.Set(); err != nil {
+			return nil, err
+		}
+
+		instance, err := ns.client.GetJivaVolume(volID)
+		if err != nil {
+			return nil, err
+		}
+
+		time.Sleep(2 * time.Second)
+		retry++
+		if instance.Status.Phase == jv.JivaVolumePhaseCreated {
```

Comment:
Is this the intention over here ? I think we should be verifying for "RW" status

Reply:
yes, volume is created means it's in RW state but to be on the safer side we can do as well